### ISSUE TITLE
Test failures with zlib-ng

### ISSUE THREAD

**Issue by tomhughes** [osmcode/osmium-tool#274]
## What version of osmium-tool are you using?

osmium version 1.16.0
libosmium version 2.20.0

## What operating system version are you using?

Fedora Rawhide (the upcoming Fedora 40 release)

## What did you do exactly?

Built osmium and ran the tests.

## What did you expect to happen?

The tests should pass.

## What did happen instead?

Tests which compare compressed PBF output failed, namely:

```
The following tests FAILED:
	328 - formats-f1xp (Failed)
	329 - formats-f1xpd (Failed)
	331 - formats-f1xpm (Failed)
	332 - formats-f1xpdm (Failed)
	335 - formats-f1pp (Failed)
	338 - formats-f1ppd (Failed)
	341 - formats-f1ppc (Failed)
Errors while running CTest
```

## What did you do to try analyzing the problem?

The problem is that Fedora 40 is [switching from zlib to zlib-ng](https://fedoraproject.org/wiki/Changes/ZlibNGTransition) for it's zlib implementation and the compressed output is different while producing the same result when decompressed.

This is in reality a broader problem in that the exact form of the compressed data could change at any time - in the past I have seen it change between zlib versions for example.

Options for fixing this would be to unpack the output before comparing it (but then you're not really testing quite the same thing) or to store multiple possible valid outputs for those tests which then requires extensions to the tooling for comparisons to support that.

**Comment by joto** [comment]
I really don't know how to best solve this.

* The tests are there to test the different formats being generated. If we change them to decode those formats, they will be testing a different thing, to the point of at least some of them being useless.
* In theory we could have a list of reference files to match against instead of just one reference file. If the test matches any of the files, the test passed. But the tests are already quite complex, because they have to be written in the horrible CMake language. I have no idea how to bring that in there.
* If we could figure out from CMake whether we have zlib or zlib-ng, we could run the checks against different reference files. Sure would be ugly, though. But I don't know how to figure that out from CMake anyway.
* We can just disable the problemtic tests in default builds and use a special CMake option to enable them.

I am leaning towards the last solution, because it is the simplest. But I am not really happy about it.

**Comment by mmd-osm** [comment]
Would unit tests also fail when using the zlib-ng-compat package instead?

**Comment by tomhughes** [comment]
Yes because that is just zlib-ng's zlib compatibility mode that builds libz.so with the traditional API calls - that is in fact exactly what Fedora is using as the zlib replacement.

There is also a libz-ng.so with a new native API which is in the main zlib-ng package.

**Comment by zenonp** [comment]
I just run into this exact same problem while building osmium-tool-1.16.0-4 (srpm from fedora 40) in mock on almalinux 9.3. However, almalinux AFAIK does not use zlib-ng. The mock build installed zlib-devel 1.2.11-40.el9 from the appstream repo.

**Comment by mmd-osm** [comment]
So if the issue is really due to spec files (which are maintained elsewhere), we can probably close this one.

**Comment by zenonp** [comment]
@mmd-osm: Yes, it is an rpm packaging bug, not an osmium-tool bug. But both the problem and the solution have now been documented here and will be useful to the next person facing the same problem.

**Comment by tomhughes** [comment]
It's not really a packaging bug, it's a "running the tests in any environment that doesn't match upstream" bug. 

Incidentally the suggested patch doesn't work (you can't make patch declarations conditional) and is also irrelevant to me as Fedora packager because I'm writing a spec to build for Fedora not a spec to build for arbitrary rpm based distros. Writing a spec for almalinux or RHEL or whatever is a matter for the packager of those distros,

**Comment by joto** [comment]
I don't have a good solution for this, so I implemented a bad one. There is now a CMake option which allows disabling those tests. Run CMake with `-DRUN_TESTS_WITH_BINARY_COMPARE=OFF` to disable them.
